src/main/java/duke/util/TaskList.java

```diff
@@ -0,0 +1,304 @@
+     * @param index Index of task in the list
+     * @throws DukeException if invalid task number is passed to this method
+     */
+    public String markItemComplete(int index) throws DukeException {
+        if (index <= 0 || index > count) {
+            throw new DukeException("Invalid task number!");
+        }
+        Task t = taskList.get(index - 1);
+        boolean canComplete = t.complete();
+        String result = "";
+        if (canComplete) {
+            result = "Nice! I've marked this task as done:\n";
+            result += t;
+        } else {
+            result = "Oops! This task is already completed!\n";
+            result += t;
+        }
+        return result;
+    }
+
+    /**
+     * Deletes a task item from the list.
+     * @param index Index of task in the list
+     * @return String representation of deleted task
+     * @throws DukeException if invalid task number is passed to this method
+     */
+    public String deleteItem(int index) throws DukeException {
+        if (index <= 0 || index > count) {
+            throw new DukeException("Invalid task number!");
+        }
+        index--;
+        Task t = taskList.get(index);
+        taskList.remove(index);
+        String result = "Noted. I've removed this task:\n";
+        result += t + "\n";
+        count -= 1;
+        if (count == 1) {
+            result += "Now you have 1 task in the list";
+        } else {
+            result += "Now you have " + count + " tasks in the list";
+        }
+        return result;
+    }
+
+    /**
+     * Creates a new task from a given input.
+     * @param inputParts An array of <code>String</code> split into type of task, name of task and date (if required)
+     * @throws DukeException if command is invalid
+     */
+    public String registerNewTask(String[] inputParts) throws DukeException {
+        checkCommandValidity(inputParts[0]);
+        Task t = addToList(inputParts[1], inputParts[0]);
+        return echo(t);
+    }
+
+    /**
+     * Checks if a given command is valid.
+     * @param type Type of command
+     * @throws DukeException if command is invalid
+     */
+    static void checkCommandValidity(String type) throws DukeException {
+        if (!type.equals("todo") && !type.equals("deadline") && !type.equals("event")) {
+            throw new DukeException("I don't know what that means :(");
+        }
+    }
+
+    /**
+     * Adds a task item to the tasklist.
+     * @param s name of task
+     * @param type type of task
+     * @return Task object to be appended to the tasklist
+     * @throws DukeException if description of task is empty or if format is incorrect
+     */
+    public Task addToList(String s, String type) throws DukeException {
+        String trimmed = s.replaceAll("^\\s+", "");
+        if (trimmed.equals("")) {
+            throw new DukeException("Description cannot be empty!");
+        }
+        if (type.equals("todo")) {
```

Comment:
Should all these differentiating types of tasks and splitting be left to the parser instead?

Reply:
Considering that the `Parser` object is a higher-level object, my understanding is that it should not meddle with the lower-level details like checking if a task is valid or not. Consider that I have implemented the extension for `Notes`. I would prefer to let their respective lists verify their validity rather than having a general all-purpose Parser exposed to the lower level details.